apps/src/p5lab/spritelab/commands/criterionCommands.js

```diff
@@ -36,6 +36,29 @@ export const commands = {
     return result;
   },
 
+  // Return true if the specified sprite began speaking
+  // and the text was not an empty string.
+  strictSpriteSpeechRenderedThisFrame(spriteId) {
+    return (
+      this.getLastSpeechBubbleForSpriteId(spriteId)?.renderFrame ===
+        this.currentFrame() &&
+      this.getLastSpeechBubbleForSpriteId(spriteId)?.text
+    );
+  },
+
+  // Return true if any sprite began speaking
+  // and the text was not an empty string.
+  strictAnySpriteSpeaks() {
+    const spriteIds = this.getSpriteIdsInUse();
+    let result = false;
+    for (let i = 0; i < spriteIds.length; i++) {
+      if (commands.strictSpriteSpeechRenderedThisFrame.call(this, i)) {
+        result = true;
+      }
+    }
+    return result;
```

Comment:
(same comment goes for the `for` loop in `strictAnySpriteSpeaking`) i think we can simplify this calculation with [`.filter`](https://developer.mozilla.org/en-US/docs/Web/JavaScript/Reference/Global_Objects/Array/filter). you'll see calculations like this commonly expressed as:

```javascript
spriteIds
  .filter((_spriteId, index) => commands.strictSpriteSpeechRenderedThisFrame.call(this, index))
  .length > 0;
```

the end result is the same, but it makes the code more functional (as in "functional programming," not related to functionality) and readable

Reply:
Love it! I changed both the functions you mentioned, as well as a couple more that were very similar. Reading up on how you can specify both `(element, id)` led me to realize there was actually an edge case bug that I also fixed. 

`getSpriteIds()` returns an array of ids. In most cases, the index matches the element, but if a sprite is removed, they stop. We needed to be passing in the element, not the id, to be sure we were checking the right sprites.